internal/app/server/server.go

```diff
@@ -49,52 +64,41 @@ func Run(bootstrapConfig *bootstrapv1.Bootstrap,
 		ctx,
 		upstreamAddress,
 		upstream.CallOptions{Timeout: time.Minute},
-		logger.Named("xdsclient"),
+		logger,
 	)
 	if err != nil {
 		logger.With("error", err).Panic(ctx, "failed to initialize upstream client")
 	}
-
-	// Initialize request aggregation mapper component.
-	requestMapper := mapper.NewMapper(aggregationRulesConfig)
-
 	// Initialize orchestrator.
 	orchestrator := orchestrator.New(ctx, logger, requestMapper, upstreamClient, bootstrapConfig.Cache)
-
 	// Start server.
 	gcpServer := gcp.NewServer(ctx, orchestrator, nil)
-	server := grpc.NewServer()
-	serverPort := strconv.FormatUint(uint64(bootstrapConfig.Server.Address.PortValue), 10)
-	serverAddress := net.JoinHostPort(bootstrapConfig.Server.Address.Address, serverPort)
-	listener, err := net.Listen("tcp", serverAddress) // #nosec
-	if err != nil {
-		logger.With("err", err).Fatal(ctx, "failed to bind server to listener")
-	}
-
 	api.RegisterEndpointDiscoveryServiceServer(server, gcpServer)
 	api.RegisterClusterDiscoveryServiceServer(server, gcpServer)
 	api.RegisterRouteDiscoveryServiceServer(server, gcpServer)
 	api.RegisterListenerDiscoveryServiceServer(server, gcpServer)
 
-	if mode != "serve" {
-		return
-	}
-
-	registerShutdownHandler(server)
+	registerShutdownHandler(server, logger)
 	logger.With("address", listener.Addr()).Info(ctx, "Initializing server")
 	if err := server.Serve(listener); err != nil {
 		logger.With("err", err).Fatal(ctx, "failed to initialize server")
 	}
 }
 
-func registerShutdownHandler(server *grpc.Server) {
+func registerShutdownHandler(server *grpc.Server, logger log.Logger) {
 	sigs := make(chan os.Signal, 1)
+	ctx := context.Background()
 	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
 	go func() {
-		<-sigs
-		_ = util.DoWithTimeout(context.Background(), func() error {
+		sig := <-sigs
+		logger.Info(ctx, "received interrupt signal:", sig.String())
+		err := util.DoWithTimeout(ctx, func() error {
```

Comment:
Would it be possible to pass in a Timeout via WithTimeout through the context? That way, we can add tests when the timeout was hit and we did not gracefully shutdown.

Reply:
I think passing a timeout via WithTimeout through the context could lead to bugs. Usually the `ctx` param in any function is the application context. Putting the notion of passing in a timeout context in param could lead to bugs.
Also the `DoWithTimeout` function is already unit tested for all scenarios it serves.

Since the code deals with system signals i felt it would be hard to unit test unless the system signals and grpc methods are mocked out. I felt it would lead to complicated indirections. I thought about trying to make `registerShutdownHandler` a utility function and try to mock and unit test the negative case, but i didn;t pursue it because the shutdown handler is not supposed to be a utility and not supposed to be used anywhere else in the code.

Reply:
Yeah but I think it's important to test the behavior that everything does shutdown, whether the client/server is misbehaving.

You can set the WithCancel or WithTimeout as an application variable so that the timeout is pretty short. I also figured you want the timeout to be configurable in some way.

Reply:
Yeah we should make the timeout configurable. Will do it.

About the test, I would like to brainstorm a bit of the approach you are thinking. Are you suggesting to continue using the integration test, have a client connect to xds relay and then signal xds-relay to shutdown. Even if we do these with a short timeout, there is no way to guarantee that graceful shutdown wont happen faster than that, since we dont control grpc stack.
Or are you suggesting to use unit tests to validate the behaviour?

Reply:
I have added a unit test. @lita lmk if the test will work for the scenario.